### ISSUE TITLE
explore QEMU for automated/unit testing

### ISSUE THREAD

**Issue by japaric** [japaric/cu#5]
~~It would be helpful for people that want to tag along but don't have hardware.~~ [DONE](http://japaric.github.io/copper/first/qemu.html)

Figure out how to build an automated testing system that runs the unit tests under qemu.

**Comment by polarathene** [comment]
This would be great, perhaps even for those with hardware to do some automated testing that could run in QEMU? I've read the support can be difficult to match hardware, [Zephyr supports a few ARM platforms via QEMU](https://www.zephyrproject.org/doc/board/arduino_due.html). Still new to all this, no idea if that's helpful.

**Comment by japaric** [comment]
The copper book now has a [chapter](http://japaric.github.io/copper/first/qemu.html) about using qemu. So, I'm going to re purpose this issue to track using qemu for unit testing.

> This would be great, perhaps even for those with hardware to do some automated testing

Yeah, this is something I want to explore.

> I've read the support can be difficult to match hardware

AFAIK, qemu only supports emulating two ARM Cortex-M microcontrollers. One of them appears to be undocumented (no data sheet, reference manual) AFAICT.

>  Zephyr supports a few ARM platforms via QEMU

It only supports the LM3S6965, which is the emulation target I'm also using in copper.

**Comment by polarathene** [comment]
> It only supports the LM3S6965, which is the emulation target I'm also using in copper.

Right.. I saw QEMU support above Arduino Due and the other mentioned boards thinking they were all QEMU supported, but they were about flashing Zephyr to them., my bad.
